src/adapter/base.cr

```diff
@@ -5,34 +5,21 @@ require "db"
 # objects to perform actions against a specific database.  Each adapter needs
 # to implement these methods.
 abstract class Granite::Adapter::Base
-  property database : DB::Database
+  property name : String
+  property url : String
 
-  def initialize(adapter : String)
-    if url = ENV["DATABASE_URL"]? || Granite.settings.database_url || replace_env_vars(settings(adapter)["database"].to_s)
-      @database = DB.open(url)
-    else
-      raise "database url needs to be set in the config/database.yml or DATABASE_URL environment variable"
-    end
-  end
-
-  DATABASE_YML = "config/database.yml"
-
-  def settings(adapter : String)
-    if File.exists?(DATABASE_YML) &&
-       (yaml = YAML.parse(File.read DATABASE_YML)) &&
-       (settings = yaml[adapter])
-      settings
-    else
-      return {"database": ""}
-    end
+  def initialize(connection_hash : NamedTuple(url: String, name: String))
+    raise "Adapter with name '#{connection_hash[:name]}' has already been registered." if Granite::Settings.adapters.any? { |adapter| adapter.name == connection_hash[:name] }
```

Comment:
This check logic feels out of place here, because it feels like this specific instance of an adapter shouldn't have knowledge of the rest of the adapters. Is there another place it could live where it's not crossing domain boundaries?

Reply:
Not sure where else to put it.  Can't override the `<<` method in the Settings class so this was only other place that kinda worked since it has access to name of adapter being instantiated.